### PR TITLE
DL-1834: auth upgrade, amended uris in AccountUtils, fixed tests

### PR DESCRIPTION
**Pull request by benryan1790**
# JIRA Story or Github Number - descriptive title

**Bug fix / New feature** (delete as appropriate)

Please include a summary / description of the change and which issue it fixes.  Include any relevant user needs, context or links to other PRs related to this PR (eg. acceptance tests, environment config).

## Checklist

 - [x]  I've made every effort to commit high quality, clean code and I have executed relevant static analyses to be sure
 - [x]  I've included appropriate tests with any code I've added (Unit, Integration, Acceptance etc.)
 - [x]  I've executed the acceptance test pack locally to ensure there are no functional regressions
 - [x]  I've added my code using logical, atomic commits, squashing as appropriate - including the JIRA issue number in the commit message
 - [x]  I've run a dependency check to ensure all dependencies are up to date

### CODE REVIEW

File: app/connectors/AWRSConnector.scala [hunk 1]
```diff
@@ -17,56 +17,57 @@
...
 }
 
-trait AWRSConnector extends ServicesConfig with RawResponseReads with LoggingUtils {
+trait AWRSConnector extends ServicesConfig with RawResponseReads with LoggingUtils with AccountUtils {
 
+  private final val subscriptionTypeJSPath = "subscriptionTypeFrontEnd"
   lazy val serviceURL = baseUrl("awrs")
-
+  lazy val approvedInfo: String =
+    (1 to Try(Messages("awrs.application_status.info.approved.count").toInt).toOption.getOrElse(0)).map {
```
Comment (Ufuoma-Daniel) [inline app/connectors/AWRSConnector.scala]: What is this? "awrs.application_status.info.approved.count" doesn't exist so won't $i always default to 0?

Comment (Ufuoma-Daniel) [reply app/connectors/AWRSConnector.scala]: Okay I see, this was something you moved up from below.

File: app/controllers/ApplicationDeclarationController.scala [hunk 2]
```diff
@@ -50,70 +46,76 @@ trait ApplicationDeclarationController extends AwrsController with AccountUtils
...
-                _ <- backUpSave4LaterInKeyStore
-                successResponse <- applicationService.updateApplication()
-              } yield Redirect(controllers.routes.ConfirmationController.showApplicationUpdateConfirmation(false)).addAwrsRefToSession(successResponse.etmpFormBundleNumber)
-            }.recover {
-              case error: ResubmissionException => InternalServerError(views.html.error_template(Messages("awrs.application_resubmission_error.title"), Messages("awrs.application_resubmission_error.heading"), Messages("awrs.application_resubmission_error.message")))
-              case error: DESValidationException => InternalServerError(views.html.error_template(Messages("awrs.application_des_validation.title"), Messages("awrs.application_des_validation.heading"), Messages("awrs.application_des_validation.message")))
-              case error =>
-                warn("Exception encountered in Application Declaration Controller :\n" + error)
-                throw error
+  private def getRefNo(enrolments: Set[Enrolment]): String =
```
Comment (Ufuoma-Daniel) [inline app/controllers/ApplicationDeclarationController.scala]: This function gets used in one place for the variable "awrs" which then doesn't get used anywhere.

File: app/controllers/BusinessDetailsController.scala [hunk 3]
```diff
@@ -134,4 +146,5 @@ object BusinessDetailsController extends BusinessDetailsController {
   override val authConnector = FrontendAuthConnector
   override val save4LaterService = Save4LaterService
   override val keyStoreService = KeyStoreService
+  val signInUrl = ExternalUrls.signIn
```
Comment (Ufuoma-Daniel) [inline app/controllers/BusinessDetailsController.scala]: does this really need to be added to the bottom of all the controllers?

File: app/utils/AccountUtils.scala [hunk 4]
```diff
@@ -16,68 +16,77 @@
...
   }
 
-  def getAwrsRefNo(implicit user: AuthContext): AwrsUtr = {
-    user.principal.accounts.awrs.get.utr
+  def getAwrsRefNo(enrolments: Set[Enrolment]): String = {
+    val refno: Option[String] = enrolments.collectFirst {
+      case enrolment if enrolment.key == "HMRC-AWRS-ORG" => enrolment.identifiers.find(_.key == "AWRSRefNumber").get.value
+    }
+
+    refno.get
```
Comment (Ufuoma-Daniel) [inline app/utils/AccountUtils.scala]: Not sure I'm a fan of a .get.

File: app/services/ModelUpdateService.scala [hunk 5]
```diff
@@ -16,22 +16,24 @@
...
 trait ModelUpdateService {
-  def ensureAllModelsAreUpToDate(implicit user: AuthContext, hc: HeaderCarrier, save4LaterService: Save4LaterService): Future[Boolean]
+  def ensureAllModelsAreUpToDate(authRetrievals: StandardAuthRetrievals
+                                )(implicit hc: HeaderCarrier, save4LaterService: Save4LaterService): Future[Boolean]
 }
 
 object NoUpdatesRequired extends ModelUpdateService {
-  override def ensureAllModelsAreUpToDate(implicit user: AuthContext, hc: HeaderCarrier, save4LaterService: Save4LaterService): Future[Boolean] = Future.successful(true)
+  override def ensureAllModelsAreUpToDate(authRetrievals: StandardAuthRetrievals)
+                                         (implicit hc: HeaderCarrier, save4LaterService: Save4LaterService): Future[Boolean] = Future.successful(true)
```
Comment (Ufuoma-Daniel) [inline app/services/ModelUpdateService.scala]: But why? Why have a function that just returns a future.successful(true)? Can this object be removed?

File: app/utils/AccountUtils.scala [hunk 6]
```diff
@@ -16,68 +16,77 @@
...
-  def getUtrOrName()(implicit user: AuthContext) = {
-    (user.principal.accounts.sa, user.principal.accounts.ct, user.principal.accounts.org) match {
-      case (Some(sa), _, _) => sa.utr.utr
-      case (None, Some(ct), _) => ct.utr.utr
-      case (None, None, Some(org)) => org.org.org
-      case _ => throw new RuntimeException("No data found")
+  val defaultEnrolmentSet = Set(Enrolment("HMRC-AWRS-ORG", Seq(EnrolmentIdentifier("AWRSRefNumber", "0123456")), "activated"),
+    Enrolment("IR-CT", Seq(EnrolmentIdentifier("UTR", "6543210")), "activated"), Enrolment("IR-SA", Seq(EnrolmentIdentifier("UTR", "0123456")), "activated"))
+
+  val defaultSaEnrolmentSet = Set(Enrolment("HMRC-AWRS-ORG", Seq(EnrolmentIdentifier("AWRSRefNumber", "0123456")), "activated"),
```
Comment (Ufuoma-Daniel) [inline app/utils/AccountUtils.scala]: This was made only for tests, it should not be in the trait.

File: app/utils/AccountUtils.scala [hunk 7]
```diff
@@ -16,68 +16,77 @@
...
+    Enrolment("IR-SA", Seq(EnrolmentIdentifier("UTR", "0123456")), "activated"))
+
+  val defaultAuthRetrieval = StandardAuthRetrievals(defaultEnrolmentSet, Some(AffinityGroup.Organisation))
```
Comment (Ufuoma-Daniel) [inline app/utils/AccountUtils.scala]: So was this and the 'defaultEnrolmentSet', should be moved to a test util or something.

File: app/utils/AccountUtils.scala [hunk 8]
```diff
@@ -16,68 +16,77 @@
...
   }
 
-  def authLink(implicit user: AuthContext): String = {
-    (user.principal.accounts.org, user.principal.accounts.sa, user.principal.accounts.agent) match {
-      case (Some(orgAccount), _, _) => orgAccount.link
-      case (None, Some(saAccount), _) => saAccount.link.replaceAllLiterally("/individual", "")
+  def authLink(authRetrievals: StandardAuthRetrievals): String = {
+    (authRetrievals.affinityGroup, authRetrievals.enrolments.find(_.key == "IR-SA")) match {
+      case (Some(AffinityGroup.Organisation), _) => "/org/UNUSED"
+      case (_,            Some(enrolment)) => s"/sa/${enrolment.identifiers.find(_.key == "UTR").get.value}"
```
Comment (Ufuoma-Daniel) [inline app/utils/AccountUtils.scala]: I'm unsure, is saAccount the same as an enrolment?

File: project/FrontendBuild.scala [hunk 9]
```diff
@@ -12,30 +12,30 @@ private object AppDependencies {
   import play.core.PlayVersion
   import play.sbt.PlayImport._
 
-  private val frontendbootstrap = "12.6.0"
+  private val frontendbootstrap = "12.7.0"
   private val domainVersion = "5.6.0-play-25"
-  private val hmrcTestVersion = "3.6.0-play-25"
+  private val hmrcTestVersion = "3.8.0-play-25"
```
Comment (Ufuoma-Daniel) [inline project/FrontendBuild.scala]: Are we fine to have our versions be the *-play-25 versions when we are still on frontend bootstrap?

File: test/controllers/BusinessNameChangeControllerTest.scala [hunk 10]
```diff
@@ -46,6 +47,7 @@ class BusinessNameChangeControllerTest extends AwrsUnitTestTraits
     override val save4LaterService = TestSave4LaterService
     override val keyStoreService = TestKeyStoreService
     override val indexService = mock[IndexService]
+    val signInUrl = ExternalUrls.signIn
```
Comment (Ufuoma-Daniel) [inline test/controllers/BusinessNameChangeControllerTest.scala]: Is this needed?

File: app/controllers/BusinessDetailsController.scala (hunk 3 above)
Comment (CEPoole) [reply app/controllers/BusinessDetailsController.scala]: Yes

**Comment by CEPoole** [comment]
Sorry for the conflicts ❤️

File: app/utils/AccountUtils.scala [hunk 11]
```diff
@@ -16,68 +16,69 @@
...
-      case true => Some(true)
-      case _ => {
-        user.principal.accounts.org.isDefined match {
-          case true => Some(true)
-          case _ => None
-        }
-      }
-    }
+  def isOrgAccount(authRetrievals: StandardAuthRetrievals): Option[Boolean] = {
+    Some(authRetrievals.enrolments.exists(_.key == "IR-CT") || authRetrievals.affinityGroup == AffinityGroup.Organisation)
```
Comment (Ufuoma-Daniel) [inline app/utils/AccountUtils.scala]: Comparing unrelated types, the second part of this OR is broken.

File: app/controllers/ApplicationDeclarationController.scala (hunk 2 above)
Comment (Ufuoma-Daniel) [reply app/controllers/ApplicationDeclarationController.scala]: Make it - authRetrievals.affinityGroup.contains(AffinityGroup.Organisation)

**Pull request merged**